Question 1a (1 point): Use MergedSummary you have created at the end of Project 3. Display a table showing the number of observations for each state in this data for your own chain and your benchmark chain respectively. It is ok to use multiple tables if that is more convenient for you.

In [64]:
import pandas as pd
import numpy as np

focal_df = pd.read_csv("long_sample_csv.csv")
benchmark_df = pd.read_csv("benchmark_long.csv")
print(focal_df)
print("------------------------------")
print(benchmark_df)

                   placekey            city region           date_range_start  \
0       222-222@8dy-ssr-z2k      Bayou Cane     LA  2018-01-01 00:00:00+00:00   
1       222-222@5s6-9w8-skf       Mishawaka     IN  2018-11-19 00:00:00+00:00   
2       222-222@5qs-zmp-gc5   Oklahoma City     OK  2018-03-26 00:00:00+00:00   
3       222-222@5xq-dzd-b8v     Santa Maria     CA  2018-01-15 00:00:00+00:00   
4       222-222@62k-p8d-jqf       Stoughton     MA  2018-05-14 00:00:00+00:00   
...                     ...             ...    ...                        ...   
354041  22h-222@629-38p-4jv   West Hartford     CT  2022-12-05 00:00:00+00:00   
354042  222-222@63k-b4h-cqz      Fort Wayne     IN  2022-02-21 00:00:00+00:00   
354043  22b-222@5qv-tcj-gp9         Prosper     TX  2022-09-12 00:00:00+00:00   
354044  222-223@5z4-ktq-7dv  San Bernardino     CA  2022-10-24 00:00:00+00:00   
354045  224-222@5qw-stn-s3q        Mesquite     TX  2022-03-14 00:00:00+00:00   

                   date_ran

In [65]:
BenchmarkSummary = benchmark_df.groupby(['region', 'date_range_start', 'dayofweek']).agg(
    Nunits=('placekey', 'nunique'),
    SumDailyVisits=('dailyvisits', 'sum'),
    MaxDailyVisitsPerUnit=('dailyvisits', 'max'),
    MinDailyVisitsPerUnit=('dailyvisits', 'min'),
    MedDailyVisitsPerUnit=('dailyvisits', 'median')
).reset_index()

print(BenchmarkSummary)
print("----------------")

FocalSummary = focal_df.groupby(['region', 'date_range_start', 'dayofweek']).agg(
    Nunits=('placekey', 'nunique'),
    SumDailyVisits=('dailyvisits', 'sum'),
    MaxDailyVisitsPerUnit=('dailyvisits', 'max'),
    MinDailyVisitsPerUnit=('dailyvisits', 'min'),
    MedDailyVisitsPerUnit=('dailyvisits', 'median'),
).reset_index()

print(FocalSummary)

      region           date_range_start  dayofweek  Nunits  SumDailyVisits  \
0         AL  2018-01-01 00:00:00+00:00     Friday       4              37   
1         AL  2018-01-01 00:00:00+00:00     Monday       4              28   
2         AL  2018-01-01 00:00:00+00:00   Saturday       4              94   
3         AL  2018-01-01 00:00:00+00:00     Sunday       4              71   
4         AL  2018-01-01 00:00:00+00:00   Thursday       4              31   
...      ...                        ...        ...     ...             ...   
69176     WI  2022-12-26 00:00:00+00:00   Saturday       3              48   
69177     WI  2022-12-26 00:00:00+00:00     Sunday       3               4   
69178     WI  2022-12-26 00:00:00+00:00   Thursday       3              47   
69179     WI  2022-12-26 00:00:00+00:00    Tuesday       3              50   
69180     WI  2022-12-26 00:00:00+00:00  Wednesday       3              31   

       MaxDailyVisitsPerUnit  MinDailyVisitsPerUnit  MedDailyVi

In [66]:
#merging into merged summary

BenchmarkSummary['date_range_start'] = pd.to_datetime(BenchmarkSummary['date_range_start'])
FocalSummary['date_range_start'] = pd.to_datetime(FocalSummary['date_range_start'])

MergedSummary = pd.merge(BenchmarkSummary, FocalSummary, on=['region', 'date_range_start', 'dayofweek'], suffixes=('_Benchmark', '_Focal'), how='inner')
MergedSummary

region          date_range_start  dayofweek  Nunits_Benchmark  \
0         AL 2018-01-01 00:00:00+00:00     Friday                 4   
1         AL 2018-01-01 00:00:00+00:00     Monday                 4   
2         AL 2018-01-01 00:00:00+00:00   Saturday                 4   
3         AL 2018-01-01 00:00:00+00:00     Sunday                 4   
4         AL 2018-01-01 00:00:00+00:00   Thursday                 4   
...      ...                       ...        ...               ...   
62988     WI 2022-12-26 00:00:00+00:00   Saturday                 3   
62989     WI 2022-12-26 00:00:00+00:00     Sunday                 3   
62990     WI 2022-12-26 00:00:00+00:00   Thursday                 3   
62991     WI 2022-12-26 00:00:00+00:00    Tuesday                 3   
62992     WI 2022-12-26 00:00:00+00:00  Wednesday                 3   

       SumDailyVisits_Benchmark  MaxDailyVisitsPerUnit_Benchmark  \
0                            37                               19   
1                            28                               15   
2                            94                               39   
3                            71                               31   
4                            31                               11   
...                         ...                              ...   
62988                        48                               32   
62989                         4                                4   
62990                        47                               26   
62991                        50                               29   
62992                        31                               19   

       MinDailyVisitsPerUnit_Benchmark  MedDailyVisitsPerUnit_Benchmark  \
0                                    5                              6.5   
1                                    3                              5.0   
2                                   13                             21.0   
3                                    8                             16.0   
4                                    5                              7.5   
...                                ...                              ...   
62988                                6                             10.0   
62989                                0                              0.0   
62990                                6                             15.0   
62991                               10                             11.0   
62992                                3                              9.0   

       Nunits_Focal  SumDailyVisits_Focal  MaxDailyVisitsPerUnit_Focal  \
0                 2                    84                           50   
1                 2                    38                           25   
2                 2                    93                           48   
3                 2                    89                           51   
4                 2                    45                           24   
...             ...                   ...                          ...   
62988             1                    38                           38   
62989             1                     3                            3   
62990             1                    19                           19   
62991             1                    27                           27   
62992             1                    33                           33   

       MinDailyVisitsPerUnit_Focal  MedDailyVisitsPerUnit_Focal  
0                               34                         42.0  
1                               13                         19.0  
2                               45                         46.5  
3                               38                         44.5  
4                               21                         22.5  
...                            ...                          ...  
62988                           38                         38.0  
62989             

In [67]:
# Summing up the number of units for Focal and Benchmark by state
state_unit_counts = MergedSummary.groupby('region')[['Nunits_Focal', 'Nunits_Benchmark']].sum().reset_index()

print(state_unit_counts)

   region  Nunits_Focal  Nunits_Benchmark
0      AL          3654              7294
1      AZ         10626              7273
2      CA         42784             21924
3      CO          7308             10962
4      CT          1827              1827
5      FL         20440             17836
6      GA         18263              7308
7      ID          1827              1827
8      IL          9100             12677
9      IN          9135              9128
10     KS          1827              1827
11     KY          5481              5460
12     MD          7308             10773
13     MI          7308             14525
14     MN          7308              3654
15     MO         12789              3654
16     MS          1827              3647
17     NC          9135              9128
18     ND          3654              1827
19     NJ          9058              3654
20     NM          1827              5481
21     NV          7266              3654
22     NY          8169           

Question 1b (2 points): Download the data for Census demographics. Merge your MergedSummary data with the demographic data by the variable "State".

How many observations do you have in MergedSummary? How many observations do you have in the Census demographic data? How many observations are matched between the two after the merge? How many observations are in MergedSummary but not the demographic data? How many observations are in the demographic data but not in MergedSummary? What explains the imperfect match?

Keep only the matched observation for the rest of the project, let us call the resulting dataset as the final analysis sample.

In [77]:
census_data =  pd.read_csv("census_state_2016.csv")
census_data

#number of observations in MergedSummary
number_of_observations_MS = len(MergedSummary)
print("Number of observations: in MS", number_of_observations_MS)

#number of observations in census demographics
number_of_observations_CD = len(census_data)
print("Number of observations in Census Data" , number_of_observations_CD)
#print(census_data.columns)

MergedSummary.rename(columns={'region': 'state'}, inplace=True)
MergedSummary.rename(columns={'state': 'state_abbr'}, inplace=True)
#print(MergedSummary.columns)


Number of observations: in MS 62993
Number of observations in Census Data 52


In [79]:
merged_data = pd.merge(MergedSummary, census_data, on='state_abbr', how='outer', indicator=True)
matched_observations = len(merged_data[merged_data['_merge'] == 'both'])
ms_not_in_cd = len(merged_data[merged_data['_merge'] == 'left_only'])
cd_not_in_ms = len(merged_data[merged_data['_merge'] == 'right_only'])

print("Number of matched observations:", matched_observations)
print("Observations in MergedSummary not in Census Data:", ms_not_in_cd)
print("Observations in Census Data not in MergedSummary:", cd_not_in_ms)

final_analysis_sample = merged_data[merged_data['_merge'] == 'both'].drop(columns='_merge')
final_analysis_sample 


Number of matched observations: 62993
Observations in MergedSummary not in Census Data: 0
Observations in Census Data not in MergedSummary: 17


state_abbr          date_range_start  dayofweek  Nunits_Benchmark  \
0             AL 2018-01-01 00:00:00+00:00     Friday               4.0   
1             AL 2018-01-01 00:00:00+00:00     Monday               4.0   
2             AL 2018-01-01 00:00:00+00:00   Saturday               4.0   
3             AL 2018-01-01 00:00:00+00:00     Sunday               4.0   
4             AL 2018-01-01 00:00:00+00:00   Thursday               4.0   
...          ...                       ...        ...               ...   
62988         WI 2022-12-26 00:00:00+00:00   Saturday               3.0   
62989         WI 2022-12-26 00:00:00+00:00     Sunday               3.0   
62990         WI 2022-12-26 00:00:00+00:00   Thursday               3.0   
62991         WI 2022-12-26 00:00:00+00:00    Tuesday               3.0   
62992         WI 2022-12-26 00:00:00+00:00  Wednesday               3.0   

       SumDailyVisits_Benchmark  MaxDailyVisitsPerUnit_Benchmark  \
0                          37.0                             19.0   
1                          28.0                             15.0   
2                          94.0                             39.0   
3                          71.0                             31.0   
4                          31.0                             11.0   
...                         ...                              ...   
62988                      48.0                             32.0   
62989                       4.0                              4.0   
62990                      47.0                             26.0   
62991                      50.0                             29.0   
62992                      31.0                             19.0   

       MinDailyVisitsPerUnit_Benchmark  MedDailyVisitsPerUnit_Benchmark  \
0                                  5.0                              6.5   
1                                  3.0                              5.0   
2                                 13.0                             21.0   
3                                  8.0                             16.0   
4                                  5.0                              7.5   
...                                ...                              ...   
62988                              6.0                             10.0   
62989                              0.0                              0.0   
62990                              6.0                             15.0   
62991                             10.0                             11.0   
62992                              3.0                              9.0   

       Nunits_Focal  SumDailyVisits_Focal  ...    black  indian_alaska  \
0               2.0                  84.0  ...  1338854          56661   
1               2.0                  38.0  ...  1338854          56661   
2               2.0                  93.0  ...  1338854          56661   
3               2.0                  89.0  ...  1338854          56661   
4               2.0                  45.0  ...  1338854          56661   
...             ...                   ...  ...      ...            ...   
62988           1.0                  38.0  ...   428940          93580   
62989           1.0                   3.0  ...   428940          93580   
62990           1.0                  19.0  ...   428940          93580   
62991           1.0                  27.0  ...   428940          93580   
62992           1.0                  33.0  ...   428940          93580   

        asian hawaiian_pacific  other_races  hispanic_latino  housing_units  \
0       85704             3769        77234           199686        2230180   
1       85704             3769        77234           199686        2230180   
2       85704             3769        77234           199686        2230180   
3       85704             3769        77234           199686        2230180   
4       85704             3769        77234           199686        2230180   
...       ...           

Observations in MergedSummary not in Census Data: 0
Observations in Census Data not in MergedSummary: 17
The reason for the imperfect match in the number of observations in Census Data not in MergedSummary is because of the additional states/terrortories in census data (17 states to be exact - Puerto Rico, District of Columbia, etc. that are not included MergedSummary). MergedSummary only accounts for 35 states while census data accounts for 52, hence the imperfect match while merging. 

Question 1c (2 points): Now using the final analysis sample, define your dependent variable as

                  Y = [Ln(SumDailyVisits+1) of your focal chain - Ln(SumDailyVisits+1) of your benchmark chain].
If you believe you can have a better variable describing the performance of your focal chain relative to your benchmark chain, you are more than welcome to use your own dependent variable, as long as you explain why your choice is better.

Define the list of your potential right-hand variables, which should include day of week indicators (Monday, Tuesday …, Sunday), some demographics, some state attributes (e.g. whether the state is red or blue as you have defined in Project 3). Feel free to add new variables to this list, as long as you articulate a reason for their relevance.

Summarize all these variables in a table with the number of observations, mean, median, standard deviation, minimum and maximum. It is ok to use multiple tables if that is more convenient for you. If some variables are categorical and thus cannot appear in this summary table, create separate table(s) to describe the frequency of each value in such categorical variable(s).

In [80]:
#print(final_analysis_sample.columns)
final_analysis_sample['Y'] = np.log(final_analysis_sample['SumDailyVisits_Focal'] + 1) - np.log(final_analysis_sample['SumDailyVisits_Benchmark'] + 1)


In [81]:
import statsmodels.api as sm
#define list of potential right hand variables
#day of the week indicators
day_of_week_counts = final_analysis_sample['dayofweek'].value_counts()
print("Day of Week Distribution:\n", day_of_week_counts)
print("--------------------------------------------------")

#some demographics
continuous_vars = final_analysis_sample[['SumDailyVisits_Focal', 'SumDailyVisits_Benchmark', 'total_population', 'median_age', 'housing_units', 'age_over65']]
# Generating descriptive statistics
continuous_summary = continuous_vars.describe()
print("summarizing demographic data")
print(continuous_summary)
print("--------------------------------------------------")

#some state attributes whether state is blue or red
Red_Blue_States = pd.read_csv("RedBlueStates.csv")

Red_Blue_States['BlueState'] = (Red_Blue_States['Color'] == 'Blue').astype(int)

state_color_mapping = Red_Blue_States.set_index('State')['BlueState'].to_dict()

final_analysis_sample['BlueState'] = final_analysis_sample['state_abbr'].map(state_color_mapping)

if 'BlueState' in final_analysis_sample.columns:
    blue_state_counts = final_analysis_sample['BlueState'].value_counts()
     
    print("State Color Distribution:\n", blue_state_counts)
    
else:
    
    print("BlueState information not available.")

print("-----------------------------------------------------------------")
print("DESCRIPTIVE STATS FOR ALL THESE VARIABLES")
descriptive_stats = final_analysis_sample.describe()
print(descriptive_stats)

Day of Week Distribution:
 dayofweek
Friday       8999
Monday       8999
Saturday     8999
Sunday       8999
Thursday     8999
Tuesday      8999
Wednesday    8999
Name: count, dtype: int64
--------------------------------------------------
summarizing demographic data
       SumDailyVisits_Focal  SumDailyVisits_Benchmark  total_population  \
count          62993.000000              62993.000000      6.299300e+04   
mean             172.398251                 64.008715      8.381432e+06   
std              192.197485                 59.144515      7.854247e+06   
min                0.000000                  0.000000      7.579530e+05   
25%               57.000000                 20.000000      3.576452e+06   
50%              113.000000                 46.000000      6.016447e+06   
75%              211.000000                 90.000000      1.014679e+07   
max             2005.000000                723.000000      3.925002e+07   

         median_age  housing_units    age_over65  
coun

Part 2: Linear regression, LASSO and Ridge
 

Question 2a (2 points): Choose a list of variables to predict your Y. The list you choose could be a subset of the potential right-hand variables you have identified in Question 1c. I will refer to these variables as “featur.

 

Use the Python command “train_test_split()” to separate your final analysis sample into a training sample and a testsaple.

 

Use function MinMaxScalar() to standardize all the “features” you choose, in both the training and te samples.

 

How many observations are there in the training sample? How many in the test sample? Choose at least two variables to comment on the difference between the two samples. For example, are they different in the months covered? Are they different in certain demographics? Feel free to use other variables for this comparison.

In [85]:
from sklearn.model_selection import train_test_split

# Define features and target variable
#need to onehotencode days of the week because it is a string
day_dummies = pd.get_dummies(final_analysis_sample['dayofweek'], prefix='Day')
final_analysis_sample_encoded = pd.concat([final_analysis_sample, day_dummies], axis=1)

X = final_analysis_sample_encoded[['SumDailyVisits_Focal', 'SumDailyVisits_Benchmark', 'total_population', 'median_age', 'housing_units', 'age_over65', 'BlueState'] + list(day_dummies.columns)]
Y = final_analysis_sample_encoded['Y']

X = X.astype(float)
# Splitting the data into train and test sets
X_train, X_test, Y_train, Y_test = train_test_split(X, Y, test_size=0.2, random_state=42)

X_train = X_train.reset_index(drop=True)
Y_train = Y_train.reset_index(drop=True)
X_test = X_test.reset_index(drop=True)
Y_test = Y_test.reset_index(drop=True)

print(f"Number of observations in the training sample: {X_train.shape[0]}")
print(f"Number of observations in the test sample: {X_test.shape[0]}")


Number of observations in the training sample: 50394
Number of observations in the test sample: 12599


In [87]:
import warnings
from sklearn.preprocessing import MinMaxScaler

warnings.filterwarnings("ignore", category=FutureWarning, message="is_sparse is deprecated*")

scaler = MinMaxScaler()

X_train_scaled = scaler.fit_transform(X_train)


X_test_scaled = scaler.transform(X_test)
#print(X_train.dtypes)

X_train_scaled_df = pd.DataFrame(X_train_scaled, columns=X_train.columns)
X_test_scaled_df = pd.DataFrame(X_test_scaled, columns=X_test.columns)


number of observations in training sample is 50394 and number of obs in the test sample is 12599. choose atleast two variables to comment on the difference between the two samples...

In [89]:
variables_to_compare = ['total_population', 'median_age']

# Descriptive statistics for the training data
train_stats = X_train_scaled_df[variables_to_compare].describe()
print("Training Data Descriptive Statistics:\n", train_stats)

# Descriptive statistics for the test data
test_stats = X_test_scaled_df[variables_to_compare].describe()
print("Test Data Descriptive Statistics:\n", test_stats)


Training Data Descriptive Statistics:
        total_population    median_age
count      50394.000000  50394.000000
mean           0.197933      0.626522
std            0.204540      0.174493
min            0.000000      0.000000
25%            0.073223      0.526316
50%            0.136612      0.631579
75%            0.243916      0.728070
max            1.000000      1.000000
Test Data Descriptive Statistics:
        total_population    median_age
count      12599.000000  12599.000000
mean           0.198535      0.626601
std            0.202079      0.175213
min            0.000000      0.000000
25%            0.073223      0.526316
50%            0.136612      0.631579
75%            0.243916      0.728070
max            1.000000      1.000000


The two variables total_population, and median_age show a very similar mean in both the test and training samples meaning that both demographic variables are evenly split and well represented in both the training and test sets. This indicates that there is not any skew or bias bewteen the two samples, which is crucial in a machine learning model -> the close split and similarity in distribution show that
random split of the data was done correctly -> any patterns that were learned by the model in the training set was reflected on the test data sample

2b) Using Y and the standardized features in Question 2a, run a linear regression in your training sample. Report the coefficient estimates and R-squares.
 

What “features” have statistically significant coefficients (with p-value<0.05)? Among these significant “features”, which has a positive coefficient and which has a negative coefficient? Comment on whether their signs and magnitude confirm your pr


 

If you extend the model to predict Y in the test sample, what is the R-squares in the testmple?

 

In [90]:
import statsmodels.api as sm

X_train_sm = sm.add_constant(X_train_scaled_df)

model = sm.OLS(Y_train, X_train_sm).fit()

# Print the summary of the regression
print(model.summary())

                            OLS Regression Results                            
Dep. Variable:                      Y   R-squared:                       0.573
Model:                            OLS   Adj. R-squared:                  0.573
Method:                 Least Squares   F-statistic:                     5211.
Date:                Sat, 20 Apr 2024   Prob (F-statistic):               0.00
Time:                        10:55:50   Log-Likelihood:                -40095.
No. Observations:               50394   AIC:                         8.022e+04
Df Residuals:                   50380   BIC:                         8.034e+04
Df Model:                          13                                         
Covariance Type:            nonrobust                                         
                               coef    std err          t      P>|t|      [0.025      0.975]
--------------------------------------------------------------------------------------------
const                   

R^2 value is 0.573 meaning that approx 57.3% of variance in Y is due to model's X inputs. Statistically signficant features p < 0.05 are SumDailyVisits_Benchmark, total_population,median_age, housing_units, and BlueState, their signs are negative, and confirm that they are statsitcally signficant. implies that as benchmark chain visits increase, performance Y decreases which matches how Y is defined (Focal - Benchmark). It also shows a negative coefficent for total population which could imply that in areas with larger populations there could be increased variety of different consumer preferences, hence focal chain has a possibility of not performing as well as the benchmark age. there is a similar negative coefficent in median_age, housing_units, and blue state show that focal chain is less poopular among certain demographics than benchmark chain, but other confounding variables can influence this. 

In [91]:
from sklearn.metrics import r2_score
X_test_sm = sm.add_constant(X_test_scaled_df)

# Predict Y using the test data
Y_pred_test = model.predict(X_test_sm)

# Calculate the R-squared for the test sample
r_squared_test = r2_score(Y_test, Y_pred_test)
print(f"R-squared for the test sample: {r_squared_test}")

R-squared for the test sample: 0.5704099441986961


Question 2c (3 points): Using Y and the standardized features in Question 2a, run LASSO in your training sample.
 

Try at least three different values for the penalty parameter. For each value you have tried, report the coefficient estimates and R-squares; also extend the model to predict Y in the test sample, and report the R-squares in the test samp
m
 

Among the penalty parameters you have tried, which one is most appropriate? Explain how you arrive at this ice.

 

In the LASSO model with your preferred choice of the penalty parameter, what “features” have zero (or very close to zero) coefficients? What and how many “features” have non-zero coefficients? Comment on whether their signs and magnitude confirm your prior. What is the R-squares when you apply this model to the test sample? Is it higher or lower than what you find in Question 2b for linear ression?

 

Use the random permutation method to identify the five most important features in the above LASSO model. What are they? Comment on whether the order of their importance is consistent wh your prior.

 

In [92]:
#running LASSO
from sklearn.linear_model import Lasso

alphas = [0.001, 0.01, 0.1]
lasso_answers = {}

for alpha in alphas:
    
    lasso_model = Lasso(alpha=alpha, random_state=42)
    
    
    lasso_model.fit(X_train_scaled, Y_train)
    
    
    Y_train_pred = lasso_model.predict(X_train_scaled)
    train_r_squared = r2_score(Y_train, Y_train_pred)
    
    
    Y_test_pred = lasso_model.predict(X_test_scaled)
    test_r_squared = r2_score(Y_test, Y_test_pred)
    
    
    lasso_answers[alpha] = {
        'coefficients': lasso_model.coef_,
        'train_r_squared': train_r_squared,
        'test_r_squared': test_r_squared
    }
    print(f"Results for alpha={alpha}:")
    print(f"Coefficients: {lasso_model.coef_}")
    print(f"Train R-squared: {train_r_squared}")
    print(f"Test R-squared: {test_r_squared}\n")

Results for alpha=0.001:
Coefficients: [ 8.20646807e+00 -8.56063630e+00 -1.79200698e+00 -5.81679204e-01
 -1.68631505e+00  3.14439095e+00 -9.86531038e-02  9.60019866e-03
 -2.90857003e-02  4.35060086e-02  4.41324908e-02 -7.52745426e-03
 -0.00000000e+00 -0.00000000e+00]
Train R-squared: 0.5609841500226345
Test R-squared: 0.5579924703758807

Results for alpha=0.01:
Coefficients: [ 3.68598578 -3.93888988 -0.         -0.46053336 -0.         -0.
 -0.15572482 -0.         -0.          0.          0.         -0.
  0.          0.        ]
Train R-squared: 0.38967633167910665
Test R-squared: 0.3948203168192118

Results for alpha=0.1:
Coefficients: [ 0. -0.  0. -0. -0. -0. -0. -0. -0. -0.  0.  0.  0.  0.]
Train R-squared: 0.0
Test R-squared: -0.0001228637874834959



Amongst the penalty parameters i've tried the most appropriate penalty parameter is alpha = 0.001 as maintains complexity of the model and has a R^2 value on the test sample that is close to the OLS regression model in question 2b)

Features that have 0 or very close to 0 coefficents are the coefficents that represent my days of the week parameters (coefficents 8 to 14... Day_Monday....Day_Sunday), etc. The first seven coefficents are non zero (SumDailyVisits_Focal, SumDailyVisits_Benchmark, total_population, median_age, housing_units,age_over65,BlueState)This matches with my OLS regression results. The R^2 coefficents are Train R-squared: 0.5609841500226345
Test R-squared: 0.5579924703758807R-squared in OLS part b): 0.573
Its is lower than what we found in Question 2b for linear regression.

Use the random permutation method to identify the five most important features in the above LASSO model. What are they? Comment on whether the order of their importance is consistent wh your prior.3


In [95]:
from sklearn.inspection import permutation_importance


best_alpha = 0.001 
lasso_best = Lasso(alpha=best_alpha)
lasso_best.fit(X_train_scaled, Y_train)


perm_importance = permutation_importance(lasso_best, X_test_scaled, Y_test, n_repeats=30, random_state=42)

# features sorted by importance
sorted_indices = perm_importance.importances_mean.argsort()

# names of the top 5 most important features
top5_feature_indices = sorted_indices[-5:]
top5_feature_names = X.columns[top5_feature_indices]


In [96]:
print("Five most important features according to permutation importance:")
for name in top5_feature_names:
    print(name)


Five most important features according to permutation importance:
total_population
housing_units
age_over65
SumDailyVisits_Benchmark
SumDailyVisits_Focal


Importance is consistent with prior because all these features above have non zero coefficents which show their effect on the Y dependent variable, this effect is consistent with the OLS regression model.

Question 2d (3 points): Using Y and the standardized features in Question 2a, run Ridge in your training sample.

Try at least three different values for the penalty parameter. For each value you have tried, report the coefficient estimates and R-squares; also extend the model to predict Y in the test sample, and report the R-squares in the test sample.

Among the penalty parameters you have tried, which one is most appropriate? Explain how you arrive at this choice.

In the Ridge model with your preferred choice of the penalty parameter, what “features” have zero (or very close to zero) coefficients? What and how many “features” have non-zero coefficients? Comment on whether their signs and magnitude confirm your prior. What is the R-squares when you apply this model to the test sample? Is it higher or lower than what you find in question 3b for linear regression and question 3c for LASSO?

Use the random permutation method to identify the five most important features in the above Ridge model. What are they? Comment on whether the order of their importance is consistent with your prior. Does Ridge identify the same five most important features as LASSO?

In [97]:
from sklearn.linear_model import Ridge

alphas = [0.001, 0.01, 0.1]

ridge_results = {}

for alpha in alphas:
    
    ridge_model = Ridge(alpha=alpha, random_state=42)
    
   
    ridge_model.fit(X_train_scaled, Y_train)
    
   
    Y_train_pred_ridge = ridge_model.predict(X_train_scaled)
    train_r_squared_ridge = r2_score(Y_train, Y_train_pred_ridge)
    
    
    Y_test_pred_ridge = ridge_model.predict(X_test_scaled)
    test_r_squared_ridge = r2_score(Y_test, Y_test_pred_ridge)
    
    
    ridge_results[alpha] = {
        'coefficients': ridge_model.coef_,
        'train_r_squared': train_r_squared_ridge,
        'test_r_squared': test_r_squared_ridge
    }
    
    print(f"Ridge Results for alpha={alpha}:")
    print(f"Coefficients: {ridge_model.coef_}")
    print(f"Train R-squared: {train_r_squared_ridge}")
    print(f"Test R-squared: {test_r_squared_ridge}\n")
    

Ridge Results for alpha=0.001:
Coefficients: [ 9.15989763e+00 -9.02157160e+00 -2.43326040e+00 -8.69227352e-01
 -4.51635346e+00  6.47631458e+00 -5.61311279e-02  2.41634453e-03
 -3.57932153e-02  2.85767328e-02  3.22973891e-02 -1.83190969e-02
 -4.97534836e-03 -4.20280907e-03]
Train R-squared: 0.5734989837250057
Test R-squared: 0.5704101507296416

Ridge Results for alpha=0.01:
Coefficients: [ 9.15863669e+00 -9.02084051e+00 -2.43389586e+00 -8.69124337e-01
 -4.51231781e+00  6.47299539e+00 -5.61566393e-02  2.42060715e-03
 -3.58028818e-02  2.85894855e-02  3.23091442e-02 -1.83235598e-02
 -4.98356577e-03 -4.20922971e-03]
Train R-squared: 0.5734989654661599
Test R-squared: 0.5704119925995126

Ridge Results for alpha=0.1:
Coefficients: [ 9.14606524e+00 -9.01351451e+00 -2.43995814e+00 -8.68085270e-01
 -4.47246939e+00  6.44002970e+00 -5.64144243e-02  2.46217952e-03
 -3.58986125e-02  2.87150915e-02  3.24258762e-02 -1.83674785e-02
 -5.06457238e-03 -4.27248387e-03]
Train R-squared: 0.5734971588020779
T

The most appropriate penality parameter is alpha = 0.1 as it shows a very very slight improvement in R^2 value from alpha = 0.001 and alpha = 0.1, showing that it creates the best generalization on unseen test data given patterns learnt from the training data, even though the changes are very small. Also it is closest to OLS regression value found prior .573.

Features with alpha = 0.1 that have zero coefficents are consistent with that of the Lasso Model, the days of the week coefficents (coefficents 8 to 14... Day_Monday....Day_Sunday), etc. The first seven coefficents are non zero (SumDailyVisits_Focal, SumDailyVisits_Benchmark, total_population, median_age, housing_units,age_over65,BlueState) which matches with my OLS regression results. The R^2 coefficents are Train R-squared: 0.5734971588020779
Test R-squared: 0.5704287566057429
R-squared in OLS part b): 0.573
R^2 of the test sample in Ridge is higher than the R^2 value found in Lasso but it is still very slightly lower than what we found in Question 2b for linear regression.

In [98]:
from sklearn.inspection import permutation_importance

ridge_best = Ridge(alpha=0.1)
ridge_best.fit(X_train_scaled, Y_train)

perm_importance_ridge = permutation_importance(ridge_best, X_test_scaled, Y_test, n_repeats=30, random_state=42)


important_indices = perm_importance_ridge.importances_mean.argsort()

# The five most important features
top_5_features = X.columns[important_indices[-5:]]

In [100]:
print("Five most important features in the Ridge model are:")
for name in top_5_features:
    print(name)


Five most important features in the Ridge model are:
total_population
SumDailyVisits_Benchmark
SumDailyVisits_Focal
housing_units
age_over65


Although Ridge identifies the same 5 most important features as Lasso, its order of importnace is a little different. They both identify total_population as feature with the most orderof importance (consistent between both models) However, Lasso identifies housing_units and age_over 65 with the second and third most order of importance while Ridge identifies 'SumDailyVisits_Benchmark', 'SumDailyVisits_Focal' as the second and third most order of importance. For Lasso 4th and 5th order of importance is 'SumDailyVisits_Benchmark' and 'SumDailyVisits_Focal' while in the Ridge model the 4th and 5th order of importance is housing_units and age_over65.  Even though both models have the same 5 features as the 5 that have the most order of importnace, the magintude at which each feature affects 5 is different. 